libctru/include/3ds/services/fspxi.h

```diff
@@ -0,0 +1,645 @@
+ * @param programId ID of the program.
+ * @param header Pointer to output the legacy sub banner data to.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(GetLegacySubBannerData, u32 bannerSize, FS_MediaType mediaType, u64 programId, u8* banner);
+
+/// Unknown command 47
+DEFINE_PXIFS_SERVICE_METHOD(0x47, void* buf, u32 size);
+
+/**
+ * @brief Gets the last modified time of a file in an archive.
+ * @param archive The archive that contains the file.
+ * @param out The pointer to write the timestamp to.
+ * @param path The UTF-16 path of the file.
+ * @param size The size of the path.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(GetFileLastModified, FSPXI_Archive archive, u64* out, u16* path, u32 size);
+
+/**
+ * @brief Reads from a special file.
+ * @param bytesRead Pointer to output the number of bytes read to.
+ * @param fileOffset Offset of the file.
+ * @param size Size of the buffer.
+ * @param data Buffer to read to.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(ReadSpecialFile, u32* bytesRead, u64 fileOffset, u32 size, u8* data);
+
+/**
+ * @brief Gets the size of a special file.
+ * @param fileSize Pointer to output the size to.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(GetSpecialFileSize, u64* fileSize);
+
+/**
+ * @brief Initiates a device move as the source device.
+ * @param context Pointer to output the context to.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(StartDeviceMoveAsSource, FS_DeviceMoveContext* context);
+
+/**
+ * @brief Initiates a device move as the destination device.
+ * @param context Context to use.
+ * @param clear Whether to clear the device's data first.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(StartDeviceMoveAsDestination, FS_DeviceMoveContext context, bool clear);
+
+/**
+ * @brief Reads data and stores SHA256 hashes of blocks
+ * @param file File to read from.
+ * @param bytesRead Pointer to output the number of read bytes to.
+ * @param offset Offset to read from.
+ * @param readBuffer Pointer to store read data in.
+ * @param readBufferSize Size of readBuffer.
+ * @param hashtable Pointer to store SHA256 hashes in.
+ * @param hashtableSize Size of hashtable.
+ * @param unk Unknown. Always 0x00001000?
+ */
+DEFINE_PXIFS_SERVICE_METHOD(ReadFileSHA256, FSPXI_File file, u32* bytesRead, u64 offset, void* readBuffer, u32 readBufferSize, void* hashtable, u32 hashtableSize, u32 unk);
+
+/**
+ * @brief Assumedly writes data and stores SHA256 hashes of blocks
+ * @param file File to write to.
+ * @param bytesWritten Pointer to output the number of written bytes to.
+ * @param offset Offset to write to.
+ * @param writeBuffer Buffer to write from.
+ * @param writeBufferSize Size of writeBuffer.
+ * @param hashtable Pointer to store SHA256 hashes in.
+ * @param hashtableSize Size of hashtable
+ * @param unk1 Unknown. Might match with ReadFileSHA256's unknown?
+ * @param unk2 Unknown.
+ */
+DEFINE_PXIFS_SERVICE_METHOD(WriteFileSHA256, FSPXI_File file, u32* bytesWritten, u64 offset, void* writeBuffer, u32 writeBufferSize, void* hashtable, u32 hashtableSize, u32 unk1, u32 unk2);
```

Comment:
```suggestion
DEFINE_PXIFS_SERVICE_METHOD(WriteFileSHA256, FSPXI_File file, u32* bytesWritten, u64 offset, const void* writeBuffer, u32 writeBufferSize, void* hashtable, u32 hashtableSize, u32 unk1, u32 unk2);
```

Is hashtable input or output? Adjust const correctness accordingly.

Reply:
These two (ReadFileSHA256 and WriteFileSHA256) are a couple of the commands I understand the least. Something seems very wrong with 3dbrew's description compared to FS's usage, and I'm currently trying to figure out what the actual purpose and usage of these commands is.